docs/DEVELOPMENT-GUIDE.md

```diff
@@ -138,6 +138,20 @@ BAIL(S2N_ERR_BAD_MESSAGE);
 
 the macro will set s2n_errno correctly, as well as some useful debug strings, and return `S2N_RESULT_ERROR`.
 
+#### Crypto error handling
+
+The cryptography libraries used by s2n have their own error state which s2n needs to manage as well.
+All error handling for results from a cryptographic library must include one of the `OSSL` macros.
+There are three different ways to do this:
+
+* You can manually assemble an OSSL to S2N_RESULT macro with a GUARD macro (Preferred).
+  * ```GUARD_AS_POSIX(OSSL_POSIX(EVP_example_method_name_here()));```
+  * ```GUARD_AS_RESULT(OSSL_PTR_WITH(EVP_PKEY_new(), S2N_ERR_NOMEM));```
+  
+  This lets you use one macro pick what whether you want to check a return value or pointer and the other to select the return type for the containing method.
+* You can use a `GUARD_*_OSSL` macro directly. (This is equivalant the prior constructions but less flexible.)
```

Comment:
Is it possible to make these macros more flexible? Considering ease of use, this option seems much more attractive.

Reply:
They fundamentally cannot be flexible as they are essentially a hard-coded combination of two of the prior macros. With the prior macros you can assemble them however you need. There is a combinatoric explosion of needed macros if we were to provide `GUARD_*_OSSL` macros for every possible case. (Checking: truth, posix, `S2N_RESULT`, pointer, `OSSL_SUCCESS`; and then returning: posix, `S2N_RESULT`, ptr, ???)

So, it may make sense to build a few of the most common ones as syntactic sugar, but I think that's it.